File: frontend/src/components/benchmark/content/merge_sort/implementations/merge_sort.go
package main

type MergeSort struct{}

func (ms *MergeSort) merge(arr []int, left, mid, right int) {
    n1 := mid - left + 1
    n2 := right - mid

    L := make([]int, n1)
    R := make([]int, n2)

    for i := 0; i < n1; i++ {
        L[i] = arr[left+i]
    }
    for j := 0; j < n2; j++ {
        R[j] = arr[mid+1+j]
    }

    i, j, k := 0, 0, left

    for i < n1 && j < n2 {
        if L[i] <= R[j] {
            arr[k] = L[i]
            i++
        } else {
            arr[k] = R[j]
            j++
        }
        k++
    }

    for i < n1 {
        arr[k] = L[i]
        i++
        k++
    }

    for j < n2 {
        arr[k] = R[j]
        j++
        k++
    }
}

func (ms *MergeSort) mergeSortRecursive(arr []int, left, right int) {
    if left < right {
        mid := left + (right-left)/2
        ms.mergeSortRecursive(arr, left, mid)
        ms.mergeSortRecursive(arr, mid+1, right)
        ms.merge(arr, left, mid, right)
    }
}

func (ms *MergeSort) Sort(arr []int) {
    ms.mergeSortRecursive(arr, 0, len(arr)-1)
}

func (ms *MergeSort) SortIterative(arr []int) {
    n := len(arr)

    for size := 1; size < n; size *= 2 {
        for left := 0; left < n-size; left += 2 * size {
            mid := left + size - 1
            right := left + 2*size - 1
            if right > n-1 {
                right = n - 1
            }
            ms.merge(arr, left, mid, right)
        }
    }
}

func (ms *MergeSort) mergeInPlace(arr []int, left, mid, right int) {
    i := left
    j := mid + 1

    for i <= mid && j <= right {
        if arr[i] <= arr[j] {
            i++
        } else {
            temp := arr[j]
            for k := j; k > i; k-- {
                arr[k] = arr[k-1]
            }
            arr[i] = temp
            i++
            mid++
            j++
        }
    }
}

func (ms *MergeSort) mergeSortInPlace(arr []int, left, right int) {
    if left < right {
        mid := left + (right-left)/2
        ms.mergeSortInPlace(arr, left, mid)
        ms.mergeSortInPlace(arr, mid+1, right)
        ms.mergeInPlace(arr, left, mid, right)
    }
}

func (ms *MergeSort) SortInPlace(arr []int) {
    ms.mergeSortInPlace(arr, 0, len(arr)-1)
}

File: frontend/src/components/benchmark/content/merge_sort/implementations/merge_sort.rs
pub struct MergeSort;

impl MergeSort {
    fn merge(arr: &mut [i32], left: usize, mid: usize, right: usize) {
        let n1 = mid - left + 1;
        let n2 = right - mid;

        let mut L = vec![0; n1];
        let mut R = vec![0; n2];

        for i in 0..n1 {
            L[i] = arr[left + i];
        }
        for j in 0..n2 {
            R[j] = arr[mid + 1 + j];
        }

        let mut i = 0;
        let mut j = 0;
        let mut k = left;

        while i < n1 && j < n2 {
            if L[i] <= R[j] {
                arr[k] = L[i];
                i += 1;
            } else {
                arr[k] = R[j];
                j += 1;
            }
            k += 1;
        }

        while i < n1 {
            arr[k] = L[i];
            i += 1;
            k += 1;
        }

        while j < n2 {
            arr[k] = R[j];
            j += 1;
            k += 1;
        }
    }

    fn merge_sort_recursive(arr: &mut [i32], left: usize, right: usize) {
        if left < right {
            let mid = left + (right - left) / 2;
            Self::merge_sort_recursive(arr, left, mid);
            Self::merge_sort_recursive(arr, mid + 1, right);
            Self::merge(arr, left, mid, right);
        }
    }

    pub fn sort(arr: &mut [i32]) {
        if !arr.is_empty() {
            Self::merge_sort_recursive(arr, 0, arr.len() - 1);
        }
    }

    pub fn sort_iterative(arr: &mut [i32]) {
        let n = arr.len();
        if n < 2 {
            return;
        }

        let mut size = 1;
        while size < n {
            let mut left = 0;
            while left < n - size {
                let mid = left + size - 1;
                let right = (left + 2 * size - 1).min(n - 1);
                Self::merge(arr, left, mid, right);
                left += 2 * size;
            }
            size *= 2;
        }
    }

    fn merge_in_place(arr: &mut [i32], mut left: usize, mut mid: usize, right: usize) {
        let mut i = left;
        let mut j = mid + 1;

        while i <= mid && j <= right {
            if arr[i] <= arr[j] {
                i += 1;
            } else {
                let temp = arr[j];
                for k in (i..j).rev() {
                    arr[k + 1] = arr[k];
                }
                arr[i] = temp;
                i += 1;
                mid += 1;
                j += 1;
            }
        }
    }

    fn merge_sort_in_place(arr: &mut [i32], left: usize, right: usize) {
        if left < right {
            let mid = left + (right - left) / 2;
            Self::merge_sort_in_place(arr, left, mid);
            Self::merge_sort_in_place(arr, mid + 1, right);
            Self::merge_in_place(arr, left, mid, right);
        }
    }

    pub fn sort_in_place(arr: &mut [i32]) {
        if !arr.is_empty() {
            Self::merge_sort_in_place(arr, 0, arr.len() - 1);
        }
    }
}